3.5 release notes blog post (#136)
sourcegraph/about: blogposts/announcing-sourcegraph-3.5.md

---

> title: 'Sourcegraph 3.5: Improved searching, repository syncing, and Bitbucket Server ACLs'

Reviewer: ```suggestion
title: 'Sourcegraph 3.5: Improved searching, repository syncing, and Bitbucket Server repository permissions'
```

"ACLs" is a wonky term that we use internally and casually, but we should be consistent and call it "repository permissions".

author: Ryan Blunden and Christina Forney
publishDate: 2019-06-20T10:00-07:00
tags: [
  blog
]
slug: sourcegraph-3.5
heroImage: /sourcegraph-mark.png
published: true
---

[Sourcegraph](https://about.sourcegraph.com/) is the standard developer platform for code search and navigation at many of the largest and most exacting technology companies. With Sourcegraph, every company can get access to the same kind of tools that Google and Facebook developers use every day.

<div style="padding-left: 2rem">

[**🎄 Improved code reviews with line decorations in pull requests**](#improved-code-reviews-with-line-decorations-in-pull-requests)<br />
Use extensions like Codecov, Sentry, Datadog, and more to put info in context of a code review.

> [**🔒 Introducing Bitbucket Server ACLs**](#introducing-bitbucket-server-acls)<br />

Reviewer: ```suggestion
[**🔒 Introducing Bitbucket Server repository permissions**](#introducing-bitbucket-server-repository-permissions)<br />
```

Enforce repository permissions defined in Bitbucket Server.

[**⏎ Multi-line search with newline (`\n`) characters**](#multi-line-search-with-newline-code-classlanguage-textncode-characters)<br />

[**📅 Restrict search results to repositories with recent activity**](#restrict-search-results-to-repositories-with-recent-activity)<br />

> [**📂 Restrict search to repositories with or without a file**](#restrict-search-to-repositories-with-or-without-a-file)<br />

Reviewer: ```suggestion
[**📂 Restrict search to repositories with or without a specific filename**](#restrict-search-to-repositories-with-or-without-a-specific-filename)<br />
```

[**💾 Powering code alerts with saved searches**](#powering-code-alerts-with-saved-searches)<br />
Improved UI to separate User and Organization saved searches.

[**💡 Repository syncing status for site admins**](#repository-syncing-status-for-site-admins)<br />
Experimental feature to convey the status of cloning repositories.

[**🛠 New `orgs` field to optimize repository syncing for GitHub organizations**](#new-code-classlanguage-textorgscode-field-to-optimize-repository-syncing-for-github-organizations)<br />
Easier configuration for cloning all repos in an organization.

[**📝 Changelog**](#35-changelog)<br />

[**🎖️ Thank you**](#thank-you)<br />

</div>

**Deploy or upgrade:** [Local](https://docs.sourcegraph.com/#quickstart-guide) | [AWS](https://github.com/sourcegraph/deploy-sourcegraph-aws) | [DigitalOcean](https://marketplace.digitalocean.com/apps/sourcegraph?action=deploy&refcode=48dfb3ccb51c) | [Kubernetes cluster](https://github.com/sourcegraph/deploy-sourcegraph)

## Improved code reviews with line decorations in pull requests

> Line decorations enhance code views with Sourcegraph extensions such as [Codecov](https://sourcegraph.com/extensions/sourcegraph/codecov), [Sentry](https://sourcegraph.com/extensions/sourcegraph/sentry), and [Datadog](https://sourcegraph.com/extensions/sourcegraph/datadog-metrics). In 3.5, line decorations support is now available on pull/merge requests.

Reviewer: ```suggestion
Line decorations enhance code views with Sourcegraph extensions such as [Codecov](https://sourcegraph.com/extensions/sourcegraph/codecov), [Sentry](https://sourcegraph.com/extensions/sourcegraph/sentry), and [Datadog](https://sourcegraph.com/extensions/sourcegraph/datadog-metrics). In 3.5, line decorations are now available on pull/merge requests on GitHub, Bitbucket Server, and GitLab.
```

(Assuming this does indeed support BBS and GitLab. If not, remove "/merge".)

Here is an example of the Codecov Sourcegraph extension showing covered (and uncovered) lines in a pull request in GitHub:

![](/blog/3.5-line-decorations.png)

> ## Introducing Bitbucket Server ACLs

Reviewer: ```suggestion
## Introducing Bitbucket Server repository permissions
```

> Configuring Sourcegraph to respect Bitbucket Server ACLs for code search and navigation is now available for instances with < 2,500 repositories.

Reviewer: ```suggestion
Configuring Sourcegraph to respect Bitbucket Server repository permissions for code search and navigation is now available for instances with < 2,500 repositories.
```

```

**Note:** multi-line search is currently only supported on indexed repositories (and, inherently, default branches). We’re [working to improve this](https://github.com/sourcegraph/sourcegraph/issues/4518).

## Restrict search results to repositories with recent activity

When you’re searching over 10,000+ repositories in an enterprise setting, it’s common to get results from old repositories with no active maintainers. You can now exclude repositories with no recent commits (to their default branch) with the beta `repohascommitafter` [search filter](https://docs.sourcegraph.com/user/search/queries).

For example: Find Dockerfiles requiring upgrades from an obsolete version of Debian so you can contact the maintainers:

```
file:Dockerfile FROM debian\:jessie|8 repohascommitafter:”1 month ago” 
```

Another example: Find Terraform files with an `aws_instance` resource to analyze EBS volume configuration:

```
aws_instance lang:hcl repohascommitafter:"1 month ago"
```

**Note:** `repohascommitafter` is considered beta as performance degrades significantly on very large repository sets. We are [addressing this in 3.6](https://github.com/sourcegraph/sourcegraph/issues/4614).

> ## Restrict search to repositories with or without a file

Reviewer: ```suggestion
## Restrict search to repositories with or without a specific filename
```

An example using `repohasfile`, is finding applications with a Dockerfile, written in Go, that use Travis CI:

```
repohasfile:Dockerfile file:\.travis.yml "language: go"
```

An example of using `-repohasfile` is finding applications with a Dockerfile that do not contain a `.dockerignore` file:

```
-repohasfile:\.dockerignore file:Dockerfile
```

Another, is finding Python applications using `pytest` without a pytest.ini or pytest.conf file:

```
-repohasfile:(pytest\.ini|pytest\.conf) lang:python "import pytest"
```

## Powering code alerts with saved searches

<p class="container">
<div style="padding:56.25% 0 0 0;position:relative;">
  <iframe src="https://player.vimeo.com/video/342111852?color=0CB6F4&amp;title=0&amp;byline=" style="position:absolute;top:0;left:0;width:100%;height:100%;" frameborder="0" webkitallowfullscreen="" mozallowfullscreen="" allowfullscreen=""></iframe>
</div>
<p style="text-align: center"><a href="https://vimeo.com/342111852" target="_blank">View on Vimeo</a></p>
</p>

> [Saved searches](https://docs.sourcegraph.com/user/search/saved_searches) can act as bookmarks to frequently used searches, and when combined with email notifications, can help monitor critical parts of your code, organization wide.

Reviewer: This sentence sounds awkward and run-on.

> Saving a search is faster and easier, and is now accessed from the user and organization settings page.

Reviewer: This also sounds run-on and awkward ("Saving a search is now accessed ..."). It is also not accurate to say they live in the settings pages, because that's the page with the Monaco editor.

```suggestion
User-level and organization-level saved searches are now separate. You can view and manage saved searches in the user and organization profile areas, respectively.
```

## Repository syncing status for site admins

VIDEO

<!-- <p class="container">
<div style="padding:56.25% 0 0 0;position:relative;">
  <iframe src="https://player.vimeo.com/video/342111852?color=0CB6F4&amp;title=0&amp;byline=" style="position:absolute;top:0;left:0;width:100%;height:100%;" frameborder="0" webkitallowfullscreen="" mozallowfullscreen="" allowfullscreen=""></iframe>
</div>
<p style="text-align: center"><a href="https://vimeo.com/342111852" target="_blank">View on Vimeo</a></p>
</p> -->

> We want customers to have greater visibility into the status of repository syncing and updating operations. in 3.5, we’re experimenting with an optional new status indicator in the navigation bar.

Reviewer: ```suggestion
To give more visibility into the status of repository syncing and updating operations, we added an experimental status indicator in the navigation bar.
```

To enable the repository syncing status indicator, add this to your site configuration:

```
"experimentalFeatures": { "statusIndicator": "enabled" }
``` 

Due to its experimental status, Sourcegraph instances with 500+ repositories may experience inaccurate status updates.

## New `orgs` field to optimize repository syncing for GitHub organizations

For customers using GitHub with 1,000+ repositories, Sourcegraph would sometimes exceed GitHub’s search API rate limit during syncing, resulting in an incomplete set of cloned repositories.

> To address this, a [new `orgs` field](https://docs.sourcegraph.com/admin/external_service/github#selecting-repositories-for-code-search) has been added for customers not needing to filter the list of repositories for their organization. This uses a different GitHub API not subject to rate limiting.

Reviewer: > customers not needing to filter the list of repositories for their organization

It's unclear what this means. I think it means "for organizations that want to sync all organization repositories (not just a subset)", right?

Author: Yes, will re-word

The `orgs` field is a list of GitHub organizations:

```
“orgs”: [
    “gorilla”
]
```

> If filtering the list of repositories is required, e.g. `archived:no  forked:no` the `repositoryQuery` should be used instead of `orgs`. If the result of a query entry in `repositoryQuery` exceeds 1,000 repositories, it will need to be split out over multiple entries. Please [contact support](mailto:support@sourcegraph.com) if you require assistance.

Reviewer: I missed this in planning, but this field should have been called `organizations` not `orgs`. https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/doc/dev/style_guide.md#specific-terms-in-prose This is not high-priority.

Author: Oh! Yes, good catch. I need to get these totally ingrained :)

- The `orgs` setting in [GitHub external service config](https://docs.sourcegraph.com/admin/external_service/github) allows admins to select all repositories from the specified organizations to be synced.
- A new experimental search filter `repohascommitafter:"30 days ago"` allows users to exclude stale repositories that don't contain commits (to the branch being searched over) past a specified date from their search query.
- The `authorization` setting in the [Bitbucket Server external service config](https://docs.sourcegraph.com/admin/external_service/bitbucket_server#permissions) enables Sourcegraph to enforce the repository permissions defined in Bitbucket Server.
- A new, experimental status indicator in the navigation bar allows admins to quickly see whether the configured repositories are up to date or how many are currently being updated in the background. You can enable the status indicator with the following site configuration: `"experimentalFeatures": { "statusIndicator": "enabled" }`.

### Changed

> - The saved searches UI has changed. There is now a Saved searches page in the user and organizations settings area. A saved search appears in the settings area of the user or organization it is associated with.

Reviewer: ```suggestion
- The saved searches UI has changed. There is now a saved searches page in the user and organizations settings area. A saved search appears in the settings area of the user or organization it is associated with.
```

Don't capitalize `S`

### Removed

### Fixed

- Fixed repository search patterns which contain `.*`. Previously our optimizer would ignore `.*`, which in some cases would lead to our repository search excluding some repositories from the results.
- Fixed an issue where the Phabricator native integration would be broken on recent Phabricator versions. This fix depends on v1.2 of the [Phabricator extension](https://github.com/sourcegraph/phabricator-extension).
- Fixed an issue where the "Empty repository" banner would be shown on a repository page when starting to clone a repository.
- Prevent data inconsistency on cached archives due to restarts. (#4366)
- On the /extensions page, the UI is now less ambiguous when an extension has not been activated. (#4446)

> The [changelog for this and previous releases](https://github.com/sourcegraph/sourcegraph/blob/master/CHANGELOG.md#3.5) is available on GitHub.

Reviewer: I don't think it's important to say "and previous releases". It's true, but why does it matter? It also sounds awkward ("the changelog for this ... releases"). And why say "is available on GitHub"? That sounds like filler and needlessly mentions GitHub here.

```suggestion
See the [full Sourcegraph changelog](https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/CHANGELOG.md).
```

Author: Great suggestion!

    ```
    aws_instance lang:hcl repohascommitafter:"1 month ago"
    ```

**Note:** `repohascommitafter` is considered beta as performance degrades significantly on very large repository sets. We are [addressing this in 3.6](https://github.com/sourcegraph/sourcegraph/issues/4614).

## Restrict search to repositories with or without a specific filename

Sometimes, you want to search repositories based on the existence or absence of a file. This is now possible with the new `repohasfile` [search filter](https://docs.sourcegraph.com/user/search/queries) which comes in two forms, `repohasfile` and `-repohasfile`.

Examples:

- Find applications with a Dockerfile, written in Go, that use Travis CI:

    ```
    repohasfile:Dockerfile file:\.travis.yml "language: go"
    ```

- Find applications with a Dockerfile that do not contain a `.dockerignore` file:

    ```
    -repohasfile:\.dockerignore file:Dockerfile
    ```

- Find Python applications using `pytest` without a pytest.ini or pytest.conf file:

    ```
    -repohasfile:(pytest\.ini|pytest\.conf) lang:python "import pytest"
    ```

## Introducing Bitbucket Server ACLs

> Configuring Sourcegraph to enforce Bitbucket Server ACLs for code search and navigation is now available for instances with < 2,500 repositories.

Reviewer 2: ```suggestion

_Note: [GitHub](https://docs.sourcegraph.com/admin/repo/permissions#github) and [GitLab](https://docs.sourcegraph.com/admin/repo/permissions#gitlab) repository permissions are already supported._

Configuring Sourcegraph to enforce Bitbucket Server ACLs for code search and navigation is now available for instances with < 2,500 repositories.
```

![](/blog/3.5-org-settings.png)

![](/blog/3.5-org-repos.png)

## Repository syncing status for site admins

<p class="container">
<div style="padding:56.25% 0 0 0;position:relative;">
  <iframe src="https://player.vimeo.com/video/343706613?color=0CB6F4&amp;title=0&amp;byline=" style="position:absolute;top:0;left:0;width:100%;height:100%;" frameborder="0" webkitallowfullscreen="" mozallowfullscreen="" allowfullscreen=""></iframe>
</div>
<p style="text-align: center"><a href="https://vimeo.com/343706613" target="_blank">View on Vimeo</a></p>
</p>

We want customers to have greater visibility into the status of repository syncing and updating operations. in 3.5, we’re experimenting with an optional new status indicator in the navigation bar.

To enable the repository syncing status indicator, add this to your site configuration:

```
"experimentalFeatures": { "statusIndicator": "enabled" }
``` 

> Due to its experimental status, Sourcegraph instances with 500+ repositories may experience inaccurate status updates.

Reviewer 2: Is "inaccurate" right here? I thought it was more an issue of it being "confusing" or "not useful"?

Author: Good catch! Updating...

---
title: 'Sourcegraph 3.5: Improved searching, repository syncing, and Bitbucket Server ACLs'
author: Ryan Blunden and Christina Forney
publishDate: 2019-06-20T10:00-07:00
tags: [
  blog
]
slug: sourcegraph-3.5
heroImage: /sourcegraph-mark.png
published: true
---

[Sourcegraph](https://about.sourcegraph.com/) is the standard developer platform for code search and navigation at many of the largest and most exacting technology companies. With Sourcegraph, every company can get access to the same kind of tools that Google and Facebook developers use every day.

<div style="padding-left: 2rem">

[**🔒 Introducing Bitbucket Server ACLs**](#introducing-bitbucket-server-acls)<br />

> Enforce repository permissions defined in BitBucket Server.

Reviewer 3: ```suggestion
Enforce repository permissions defined in Bitbucket Server.
```

Easier configuration for cloning all repos in an organization.

[**📝 Changelog**](#35-changelog)<br />

[**🎖️ Thank you**](#thank-you)<br />

</div>

**Deploy or upgrade:** [Local](https://docs.sourcegraph.com/#quickstart-guide) | [AWS](https://github.com/sourcegraph/deploy-sourcegraph-aws) | [DigitalOcean](https://marketplace.digitalocean.com/apps/sourcegraph?action=deploy&refcode=48dfb3ccb51c) | [Kubernetes cluster](https://github.com/sourcegraph/deploy-sourcegraph)

## Introducing Bitbucket Server ACLs

Configuring Sourcegraph to respect Bitbucket Server ACLs for code search and navigation is now available for instances with < 2,500 repositories.

Making this possible, is the new `authorization` field, which enables Sourcegraph to communicate with the Bitbucket Server through an application link. Check out the [configuring Bitbucket Server and Sourcegraph guide](https://docs.sourcegraph.com/admin/repo/permissions#bitbucket-server), which documents the requirements and configuration process from start to finish.

Work [will continue in 3.6](#) to improve performance to scale beyond 10,000+ repositories.

## Multi-line search with newline (`\n`) characters

> Sourcegraph now supports multi-line searches by matching on `\n`, providing new search use cases. For example, to identify usage of Python decorators at the module, and class level:

Reviewer 4: ```suggestion
Sourcegraph now supports multi-line searches by matching on `\n`, providing new search use cases. For example, to identify usage of Python decorators at the module and class level:
```

An example using `repohasfile`, is finding applications with a Dockerfile, written in Go, that use Travis CI:

```
repohasfile:Dockerfile file:\.travis.yml "language: go"
```

An example of using `-repohasfile` is finding applications with a Dockerfile that do not contain a `.dockerignore` file:

```
-repohasfile:\.dockerignore file:Dockerfile
```

Another, is finding Python applications using `pytest` without a pytest.ini or pytest.conf file:

```
-repohasfile:(pytest\.ini|pytest\.conf) lang:python "import pytest"
```

## Improved code reviews with line decorations in pull requests

<!-- <p class="container">
<div style="padding:56.25% 0 0 0;position:relative;">
  <iframe src="https://player.vimeo.com/video/52c9a6daa8?color=0CB6F4&amp;title=0&amp;byline=" style="position:absolute;top:0;left:0;width:100%;height:100%;" frameborder="0" webkitallowfullscreen="" mozallowfullscreen="" allowfullscreen=""></iframe>
</div>
<p style="text-align: center"><a href="https://vimeo.com/52c9a6daa8" target="_blank">View on Vimeo</a></p>
</p> -->

CODECOV VIDEO?

> Line decorations in file views are one of the awesome features of the Sourcegraph browser extension. Now, these same line decorations are inside your pull requests on all code hosts! 

Reviewer 4: ```suggestion
Line decorations in code views are used by Sourcegraph extensions such as [Codecov](https://sourcegraph.com/extensions/sourcegraph/codecov), [Sentry](https://sourcegraph.com/extensions/sourcegraph/sentry), and [Datadog](https://sourcegraph.com/extensions/sourcegraph/datadog-metrics) and in 3.5, line decorations support is now available on pull/merge requests.
```

Reviewer 4: I'm not digging the "awesome" part. See what you think, just a suggestio.
